Texture based GPU morphing

mvaligursky opened the pull request:
Implementation of mesh morphing based on storing morph targets in float textures, as opposed to storing them in vertex attributes.

![final](https://user-images.githubusercontent.com/59932779/83641034-de3d4b80-a5a4-11ea-9831-25ed03a41af0.gif)

**Implementation details:**
- Supported platform requirements:
    - webgl 1 or 2
    - float / half float texture support
    -  float / half float texture blending support
    - float / half float render target support
    - highp fragment shader precision
    - at least 2 vertex shader textures
- this allows "unlimited" number of active morph targets per mesh (up to 4096x4096 texture size to store morph targets)
- texture based implementation is the preferred option if supported by hardware (half-float, then float precision), otherwise attribute based morphing is used.

**Future work:**
- allow callback during Morph creation by glb-parser and let user specify which render technique should be used (attribute vs texture based) and precision (float / half float)
- create cpu based fallback if more than 8 active targets are needed and texture based morphing is not supported

**Performance:**
- memory: texture based morphing is storing morph targets as partially sparse (so less memory here), but has additional render target texture for position / normal allocated (so additional memory here)
- performance (2019 Mac and Google Pixel 3A, GPU time in ms, 100 morph high res meshes, no shadows)
![Screen Shot 2020-06-05 at 2 29 43 PM](https://user-images.githubusercontent.com/59932779/83882598-42dddf00-a73a-11ea-877e-3ab7ed692035.png)
- texture based morph target blending takes place only one time per frame and so when rendering the same mesh multiple times (to shadow maps ..), performance of texture based morphing is likely to be even better than current test

**Engine example (when released):**
* show 10 morph targets modifying position and normal of a single mesh {@link http://playcanvas.github.io/#graphics/mesh-morph-many.html}

mvaligursky commented on src/graphics/program-lib/chunks/transform.vert:
```diff
@@ -57,6 +78,13 @@ vec4 getPosition() {
 
+    #ifdef MORPHING_TEXTURE_BASED_POSITION
+        // apply morph offset from texture
+        vec2 morphUV = getTextureMorphCoords();
```
shader compiler will easily remove the duplicate code, that's the main thing they do when they vectorize the shader - so I picked nicer solution :)

mvaligursky commented on src/scene/morph.js:
```diff
@@ -29,19 +39,146 @@ Object.assign(pc, function () {
+
+            // if texture morphing is not set up, use attribute based morphing
+            var i;
+            if (!this._useTextureMorph) {
```
that would change the behavior. Code above (texture based init) can fail and set _useTextureMorph to false .. so sometimes both paths execute.

mvaligursky commented:
I considered half precision, but webglstats doesn't have show any stats on OES_texture_half_float extension so decided against it in the first release.

I created an issue with some possible follow up work:
https://github.com/playcanvas/engine/issues/2126

I think it's very useful to have attribute based backup - this way morphing is available on every device with a good performance (and some limitations on concurrent morph targets). Attribute based morphing is also faster and could be a first choice for many application needed some morph targets.

slimbuck commented:
Actually, having recently worked with RGBE format for images, I wonder whether you couldn't do exactly the same with position/normal data? You can see pack/unpack implemented at https://github.com/slimbuck/engine/blob/hdr-parser/src/graphics/program-lib/chunks/reproject.frag#L55

kenrussell commented:
There are a considerable number of mobile devices that only support half-float render targets, not full floats - including iOS devices. These devices also don't support OES_texture_float_linear. For this reason I'd recommend adding half-float support to this PR early on, ideally in the first version.

The 2x GPU time compared to vertex attribute based morphing might be influenced by the amount of data being fetched from the FP32 textures. It would be interesting to try preferentially using FP16 textures and see if it provides a speedup. Maybe users could opt-in to FP32 support (if supported on the hardware) if they find they need the additional precision?

Availability data for the underlying OpenGL ES extensions can be found at https://opengles.gpuinfo.org/listextensions.php - though it doesn't seem to paint the whole picture because it shows the FP16 extensions (EXT_color_buffer_half_float, OES_texture_half_float) as being less available than the FP32 extensions (EXT_color_buffer_float, OES_texture_float).

mvaligursky commented:
@kenrussell - thanks for the suggestion, I didn't know float render targets are not supported on iOS (don't have a device to test on just yet). I've added half float implementation and use it as a default one for now - it seems to give it a bit of an improvement compared to floats.

I've also updated performance data in the description - on my Pixel 3A phone, half float texture solution is faster than attribute based, which is great! I'll add iPhone data when I get my hands on my (within a week I hope)

slimbuck commented:
This is great @mvaligursky! Now when can we see the rgbe version?!

mvaligursky commented:
> This is great @mvaligursky! Now when can we see the rgbe version?!

we'd need a way for user to turn it on .. so the callbacks we spoke about.
Also, it would be rather costly .. blending 10 textures would mean decoding of 10 textures. And not sure I could get away with writing the result into rgbe, so might need half / float support anyways.

slimbuck commented:
Yeah cool, I'll get the callbacks in asap.

I actually doubt the coding/decoding will slow things down since that's literally all the shader is doing right? The precision of rgbe might be quite limiting though, not easy to say without trying it. Other issue is you loose target blend if that's stored rgbe.

Still, potentially halving the texture bandwidth again over float16 might be a win on the lowend devices.

Oh and also, the link to encode/decode above was just to give an idea of complexity, it's not tested and might not work properly!

kenrussell commented on src/scene/morph.js:
```diff
@@ -29,19 +49,162 @@ Object.assign(pc, function () {
+
+                // create texture and assign it to target
+                target = deltaInfos[i].target;
+                var format = this._textureFormat === pc.Morph.FORMAT_FLOAT ? pc.PIXELFORMAT_RGB32F : pc.PIXELFORMAT_RGBA16F;
```
Does this texture need to be color-renderable? RGB32F isn't per:
https://www.khronos.org/registry/OpenGL/extensions/EXT/EXT_color_buffer_float.txt

but it sounds like the code above already takes this into account, so maybe these textures are only sampled, not rendered into.

mvaligursky replied:
yes, these are only the source texture for reading, either RGB32F, or RGBA16F (RBG16F was failing for me unless I created it power of 2)

Render target texture is created here:

https://github.com/playcanvas/engine/blob/2dce02f7a89d9fde2262a2b886865995806a5cd6/src/scene/morph-instance.js#L44-L46

mvaligursky commented:
just for the reference - on iOS devices half float textures cannot be initialised with half data. This has been fixed in Oct 2019, but older devices cannot use half textures due to this.
https://bugs.webkit.org/show_bug.cgi?id=169999

conformance tests
https://www.khronos.org/registry/webgl/sdk/tests/conformance/extensions/oes-texture-half-float.html

mvaligursky commented:
closing this pull request, it's superseded by #2196

mvaligursky commented:
@kenrussell - would you know what the story is on iOS14 beta WebGl2 implementation? It seems to not support EXT_color_buffer_float and so I cannot use float nor half float framebuffers. WebGL1 implementation on the same device supports it.

kenrussell commented:
@mvaligursky thanks for pointing this out. I've had a to-do item for some time to revise the EXT_color_buffer_half_float extension to allow it to be used on WebGL 2.0 contexts. https://github.com/KhronosGroup/WebGL/issues/3093 tracks it and I've just posted a pull request for it.

Have also filed https://bugs.webkit.org/show_bug.cgi?id=216010 so that this doesn't get forgotten. We'll address this as soon as the spec change lands.
